tools/gulp/tasks/screenshots.ts

```diff
@@ -0,0 +1,190 @@
+import {task} from 'gulp';
+import {readdirSync, statSync, existsSync, mkdirSync} from 'fs';
+import {openScreenshotsCloudStorage, openFirebaseScreenshotsDatabase} from '../task_helpers';
+import * as path from 'path';
+import * as admin from 'firebase-admin';
+const request = require('request');
+const imageDiff = require('image-diff');
+
+const SCREENSHOT_DIR = './screenshots';
+const FIREBASE_FILELIST = 'screenshot/filenames';
+const FIREBASE_REPORT = 'screenshot/reports';
+
+/** Task which upload screenshots generated from e2e test. */
+task('screenshots', () => {
+  let prNumber = process.env['TRAVIS_PULL_REQUEST'];
+  if (prNumber) {
+    let database = openFirebaseScreenshotsDatabase();
+    return getScreenFilenames(database)
+      .then((filenames: string[]) => downloadAllGolds(filenames, database, prNumber))
+      .then((results: boolean) => updateResult(database, prNumber, results))
+      .then((result: boolean) => updateGithubStatus(result, prNumber))
+      .then(() => setScreenFilenames(database, prNumber))
+      .then(() => uploadScreenshots(prNumber, 'diff'))
+      .then(() => uploadScreenshots(prNumber, 'test'))
+      .then(() => updateTravis(database, prNumber))
+      .then(() => database.goOffline(), () => database.goOffline());
+  }
+});
+
+function updateFileResult(database: admin.database.Database, prNumber: string,
+                          filenameKey: string, result: boolean): admin.Promise<void>{
+  return database.ref(FIREBASE_REPORT).child(`${prNumber}/results/${filenameKey}`).set(result);
+}
+
+function updateResult(database: admin.database.Database, prNumber: string,
+                      result: boolean): admin.Promise<void> {
+  return database.ref(FIREBASE_REPORT).child(`${prNumber}/result`).set(result).then(() => result);
+}
+
+function updateTravis(database: admin.database.Database,
+                      prNumber: string): admin.Promise<void> {
+  return database.ref(FIREBASE_REPORT).child(prNumber).update({
+    'commit': process.env['TRAVIS_COMMIT'],
+    'sha': process.env['TRAVIS_PULL_REQUEST_SHA'],
+    'travis': process.env['TRAVIS_JOB_ID'],
+  });
+}
+
+/** Get a list of filenames from firebase database. */
+function getScreenFilenames(database: admin.database.Database): admin.Promise<string[]> {
+  return database.ref(FIREBASE_FILELIST).once('value')
+      .then((snapshots: admin.database.DataSnapshot) => {
+    return snapshots.val();
+  });
+}
+
+/** Upload a list of filenames to firebase database as gold. */
+function setScreenFilenames(database: admin.database.Database,
+                            reportKey?: string): admin.Promise<void> {
+  let filenames: string[] = [];
+  readdirSync(SCREENSHOT_DIR).map((file: string) => {
```

Comment:
FYI: Since you don't map anything you can just use `forEach` instead of `map.

I saw a similar logic below. How about moving this to a function?

```ts
function getLocalScreenshotFiles(): string[] {
  return readdirSync(SCREENSHOT_DIR)
    .filter(fileName => !statSync(path.join(SCREENSHOT_DIR, fileName)).isDirectory())
    .map(fileName => extractScreenshotName(fileName));
}
```